### PR TITLE
x86 : Memory domain implementation for x86

### PR DESCRIPTION
**Pull request by AdithyaBaglody**
This PR shows the implementation of memory domain for x86.

P.S.
In the PR #1516 there were some build related issues which were not caught. This PR fixes that and also reverses the reverts that happened for the PR.

Delta from PR #1516 
1. Added prototype for arm_core_mpu_remove_mem_partition in include/arch/arm/cortex_m/mpu/arm_core_mpu_dev.h

### CODE REVIEW

**Comment by andrewboie** [comment]
Needs a rebase.

Some weird test case errors too:

```
***** CPU Page Fault (error code 0x00000000)
Supervisor thread read address 0xaaaaaaaa
PDE: 0x000 Non-present Read-only Supervisor
PTE: 0x000 Non-present Read-only Supervisor
Current thread ID = 0x0041b7a4
Faulting segment:address = 0x0008:0x00008606
eax: 0xaaaaaaaa, ebx: 0x00401400, ecx: 0x00419000, edx: 0x00000001
esi: 0x00000001, edi: 0x00401460, ebp: 0x0041b764, esp: 0x0041b754
eflags: 0x86
Fatal fault in essential thread! Spinning...
Terminate emulator due to fatal kernel error
```
Not sure what's going on here, but 0xAAAAAAAA is what we initialize stack memory regions to if CONFIG_INIT_STACKS=y

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
rebase please, also need CI passing

**Comment by AdithyaBaglody** [comment]
Delta:
1. When added the x86 pae mode on top these commits there was a overflow happening with the attributes of the partition. Changed it to support different size of the attributes.
2. due to a problem in the arm build system. I am waiting on the PR #4542 to get merged.

PS shippable will complain until that is merged.

**Comment by nashif** [comment]
recheck

**Review by tyrantlizard77 — DISMISSED** [review]
You need to update sample.yaml for this test case it doesn't get build for x86.
You could remove the platform_whitelist and instead put "filter: CONFIG_ARCH_HAS_USERSPACE"

**Comment by andrewboie** [comment]
sample.yaml taken care of in #4678 
This needs to be rebased though

File: arch/arm/core/cortex_m/mpu/arm_core_mpu.c
Comment (andrewboie) [inline arch/arm/core/cortex_m/mpu/arm_core_mpu.c]: you don't need to put documentation here.
document this in the arch-independent header

File: arch/arm/core/cortex_m/mpu/arm_core_mpu.c
Comment (andrewboie) [inline arch/arm/core/cortex_m/mpu/arm_core_mpu.c]: same as above

File: kernel/include/nano_internal.h
Comment (andrewboie) [inline kernel/include/nano_internal.h]: Zephyr coding standard is to have the verb last.
Should be _arch_mem_domain_partition_remove

**Comment by AdithyaBaglody** [comment]
Delta:  Handled review comments from Andrew.
Pending: Header file inclusion sequence issue. Shippable will continue to complain

File: samples/mpu/mem_domain_apis_test/src/main.c
Comment (lpereira) [inline samples/mpu/mem_domain_apis_test/src/main.c]: Can we have a `K_MEM_PARTITION_ALIGNED` (or something of the sort) that each platform specifies with the correct page size?

Comment (andrewboie) [reply samples/mpu/mem_domain_apis_test/src/main.c]: It's hard to capture this in a completely generic way.
On x86, memory is configured at 4K page granularity, the regions have to be 4K aligned but the size can be any multiple of 4K.
With ARM MPU, a region can be of any power-of-two size starting at 32 bytes, and must be aligned to its size.
Other MPUs may be more permissive.
This sample needs to be rewritten as a test case anyway as it doesn't actually prove that anything works in terms that CI can understand.

Comment (andrewboie) [reply samples/mpu/mem_domain_apis_test/src/main.c]: Having said this, one cheap way to get rid of the ifdef is just make the size 4096, aligned to 4096, that ought to work everywhere.

**Comment by andrewboie** [comment]
Inclusion issues should now be resolved if you can rebase this.

**Comment by andrewboie** [comment]
build failing due to documentation issues

> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/doc/api/kernel_api.rst:232: WARNING: doxygengroup: Cannot find namespace "mem_domain_apis" in doxygen xml output for project "Zephyr" from directory: doxygen/xml

File: include/kernel.h
Comment (dbkinder) [inline include/kernel.h]: the doxygen group "mem_domain_apis" will not be defined if CONFIG_USERSPACE is not defined, and there are references to this doxygen group in doc/api/kernel_api.rst  that will throw an error.

**Review by dbkinder — DISMISSED** [review]

File: include/kernel.h
Comment (andrewboie) [reply include/kernel.h]: IIRC, the way we do this with other conditionally compiled subsystems is to not ifdef the API definitions in the headers, but instead just conditionally compile the implementation, so that the user gets errors at link time if they try to use something that isn't enabled.

Comment (AdithyaBaglody) [reply include/kernel.h]: I understand the issue now, I was wondering what was causing this.
@dbkinder why are we doing preprocessing when generating doc using doxygen? is there a specific reason? incidentally we are missing out on a lot of information since a lot the code would depend on a config. Can we do, "ENABLE_PREPROCESSING = y" for the doxygen?. I think that was the control for doxygen, AFAIK.

Comment (AdithyaBaglody) [reply include/kernel.h]: @andrewboie I think there might be some issue regarding that solution. I want to hide the k_mem_domain and k_mem_partition under CONFIG_USERSPACE. But the mem_domain api needs these structures to even compile. Thus causing a compilation issue and not a linker issue. I will see if there is a better way.

Comment (andrewboie) [reply include/kernel.h]: what is the point of hiding these structs?
just get rid of the ifdef!

Comment (dbkinder) [reply include/kernel.h]: ENABLE_PREPROCESSING=YES is the doxygen setting we're using (see doc/zephyr.doxyfile)

Comment (dbkinder) [reply include/kernel.h]: ENABLE_PREPROCESSING has been enabled since I've been with the project... Not sure of the impact of removing it but I can check and see how it would impact the documentation.

Comment (andrewboie) [reply include/kernel.h]: enabling preprocessing will not help you because CONFIG_USERSPACE is not on by default

@AdithyaBaglody please get rid of the ifdef. We don't care about exposing the structs. We don't conditionally compile any of the k_poll API definitions even though the implementation is selectively compiled with K_POLL

**Comment by AdithyaBaglody** [comment]
Delta:
Handled the David's comments on documentation.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**
